### ISSUE TITLE
Bug: User configured MIME Categories don't get saved

### ISSUE THREAD

**Issue by Hubcapp** [shundhammer/qdirstat#109]
Using version 1.5 packaged in Debian Buster, user added MIME Categories are not preserved after restarting qdirstat.

**Comment by shundhammer** [comment]
I just checked, and this indeed happens even with the latest Git head.

At first I thought it was the old permissions / file owner problem again (`~/.config/QDirStat/*.conf` being owned by `root` after invoking `sudo qdirstat`), but this appears to be a general problem at the moment.

Thanks for reporting this!

**Comment by shundhammer** [comment]
OK, now I am confused.

I added some debug logging to `MimeCategorizer.cpp` and tried again (with Git master), and it now it works.

Please check permission and ownership of QDirStat's config dir and files:

`ls -la ~/.config/QDirStat`

It did happen in the past that invoking the program with `sudo` changed ownership of to `root` which of course means that you can't write there anymore with our normal user account.

Since some time, QDirStat checks if it's runnig with `sudo` and in that case will attempt to change ownership of its config files to the normal user after writing them. It is possible that old leftover config files from such a `sudo qdirstat` call that did not yet have that workaround changed ownership to `root`. Changing ownership to your normal user (or just invoking `sudo qdirstat´ again with the current version) should fix that.

Can this be your problem?

**Comment by Hubcapp** [comment]
```
$ ls -la ~/.config/QDirStat
total 24
drwxr-xr-x   2 tyler tyler 4096 Jun 29 21:05 .
drwx------ 125 tyler tyler 4096 Jun 30 19:56 ..
-rw-r--r--   1 tyler tyler 2367 Jun 29 21:05 QDirStat-cleanup.conf
-rw-r--r--   1 tyler tyler 2209 Jun 29 21:05 QDirStat.conf
-rw-r--r--   1 tyler tyler  312 Jun 29 21:05 QDirStat-exclude.conf
-rw-r--r--   1 tyler tyler 2195 Jun 29 21:05 QDirStat-mime.conf
```

The owner still seems to be my normal user account.

**Comment by Hubcapp** [comment]
It's worth mentioning that other user settings (such as custom patterns under each MIME Category) _do_ get saved.

**Comment by shundhammer** [comment]
OK, that is even more confusing. Thanks for that update.

**Comment by Hubcapp** [comment]
This is probably the worst kind of bug. I rechecked if I could save a MIME Category, and now it works.

Not sure what changed, but I'm encouraged that you were able to reproduce it at least one time...

**Comment by shundhammer** [comment]
OK, found it: The application creates multiple instances of the MimeCategorizer class, and depending on shutdown order, the destructors of each one write the status of that instance back to QSettings and thus to file, so the one that writes last wins. Sometimes that's the one that was used for configuring the categories in the config dialog, sometimes it's another one, and then your settings get lost.

So this needs to become a singleton as well, and all will be fine.

This problem was introduced with the "File Details" panel which also displays the MIME category of a file; for that it uses a MimeCategorizer. It turns out that creating a new instance was a bad idea.

**Comment by Hubcapp** [comment]
Ok Awesome! :-) Glad you found it. Thanks for looking into this and resolving it so quickly. The deadline for getting an updated package into Buster before release on July 6th has probably already passed, but technically "not yet" if you wanted to try.

**Comment by shundhammer** [comment]
That would be only this one patch, but there are a whole lot more changes pending:

https://hackweek.suse.com/18/projects/qdirstat-show-unpackaged-files


Along with the new package view and some performance improvements that would be a bit too disruptive, I fear. But the fix I currently work on should apply cleanly to 1.5, so a package maintainer could use it as a patch on top of it.

**Comment by Hubcapp** [comment]
I wrote the maintainer for qdirstat in Debian to let them know about this bug, and its fix.

**Comment by shundhammer** [comment]
OK, thank you!

It is still possible that QDirStat 1.6 _with_ all those new features might be ready in that time frame, but hurrying things tends to introduce breakages.

**Comment by Hubcapp** [comment]
I'm not really involved with packaging for Debian, but as I understand it, the deadline for a fix like this was probably on June 25th, and right now it's in "hard-freeze" where only critical fixes will be considered. Is a bug where the user might spend a dozen minutes configuring MIME types in qdirstat only to lose those changes on exit a critical bug?

Possibly, but, there's lots of bugs being considered, only so much volunteer time, and only 5 days to release. I would guess that this fix won't get in...

**Comment by Hubcapp** [comment]
It doesn't affect existing configurations, and it's not a security vulnerability.

**Comment by shundhammer** [comment]
Cherry-picked the fix for the stable-1.5 branch; please test:

https://github.com/shundhammer/qdirstat/tree/stable-1.5
